### PR TITLE
Netsuite dynamic host addressing

### PR DESCRIPTION
**Pull request by merit-sno**
Problem
---
NetSuite WSDL now relies on dynamic host addresses for accessing subscriber based WSDL documents.  With NetSuite 2020.1, this becomes a requirement, and zeep appears to have a failure to recognize NetSuite's use of relative addressing.

Solution
---
Wrap zeep.transports.transport with a dynamic address manager that updates the get and post services to use subscriber assigned dynamic addressing.

### CODE REVIEW

File: netsuite/client.py [hunk 1]
```diff
@@ -83,8 +84,59 @@ def wrapper(self, *args, **kw):
...
+        kwargs.pop('wsdl_url',None)
+        
+        # Find first slash after domain name
+        idx = _wsdl.index('/', 8) + 1
+
+        # We only want the dynamic domain
+        _wsdl=_wsdl[:idx]
+        self._wsdl = _wsdl
+        
+        super(NetSuiteTransport, self).__init__(**kwargs)
```
Comment (merit-mthompson) [inline netsuite/client.py]: I believe in Python 3 this will throw an error and should maybe be:

```python
super().__init__(**kwargs)
```

Comment (merit-mthompson) [reply netsuite/client.py]: When run locally, this leads to this error:

```
Traceback (most recent call last):
  File "foo.py", line 90, in <module>
    nstOG = NetSuiteTransportOG(wsdl_url=wsdl_url)
  File "foo.py", line 81, in __init__
    super(NetSuiteTransport, self).__init__(**kwargs)
TypeError: super(type, obj): obj must be an instance or subtype of type
```

File: netsuite/client.py [hunk 2]
```diff
@@ -83,8 +84,59 @@ def wrapper(self, *args, **kw):
...
+        _wsdl = kwargs['wsdl_url']
+
+        kwargs.pop('wsdl_url',None)
+        
+        # Find first slash after domain name
+        idx = _wsdl.index('/', 8) + 1
+
+        # We only want the dynamic domain
+        _wsdl=_wsdl[:idx]
+        self._wsdl = _wsdl
```
Comment (merit-mthompson) [inline netsuite/client.py]: Another way to do this that might be a little less verbose and uses the standard URL parsing library:

```python
from urllib.parse import urlparse

class NetSuiteTransport(Transport):
    def __init__(self, wsdl_url, *args, **kwargs):
        parsed_wsdl_url = urlparse(wsdl_url)
        self._wsdl = f'{parsed_wsdl_url.scheme}://{parsed_wsdl_url.netloc}/'
        super().__init__(**kwargs)
```

File: netsuite/client.py [hunk 3]
```diff
@@ -83,8 +84,59 @@ def wrapper(self, *args, **kw):
...
+        """
+        
+        idx = address.index('/',8) + 1;
+        address = self._wsdl + address[idx:]
+        
+    def get(self, address, params, headers):
+        """
+        Update the GET address before providing it to zeep.transports.transport
+        """
+        return super(NetSuiteTransport, self).get(self._fix_address(address), params, headers)
```
Comment (merit-mthompson) [inline netsuite/client.py]: I think like the other super() calls this should maybe be `return super().get(self._fix_address(address), params, headers)`

File: netsuite/client.py [hunk 4]
```diff
@@ -83,8 +84,59 @@ def wrapper(self, *args, **kw):
...
+
+    def post(self, address, message, headers):
+        """
+        Update the POST address before providing it to zeep.transports.transport
+        """
+        return super(NetSuiteTransport, self).post(self._fix_address(address), message, headers)
```
Comment (merit-mthompson) [inline netsuite/client.py]: I think like the other super() calls this should maybe be `return super().post(self._fix_address(address), message, headers)`

File: setup.py [hunk 5]
```diff
@@ -2,13 +2,16 @@
...
+    version='0.4.1',
     description='Wrapper around Netsuite SuiteTalk Web Services and Restlets',
     packages=['netsuite'],
     include_package_data=True,
     author='Jacob Magnusson',
     author_email='m@jacobian.se',
-    url='https://github.com/jmagnusson/netsuite',
+    fork_author='Stephen Opal'
+    fork_email='sno@merit.edu'
+    fork_org='Merit Network, Inc.'
```
Comment (merit-mthompson) [inline setup.py]: I love the intent here, but this dictionary will be passed/used elsewhere and invalid kwargs aren't ideal. Can we switch these to be our own, clobbering the original author's so we don't add unexpected kwargs?

**Review by merit-mthompson — CHANGES_REQUESTED** [review]

**Comment by merit-mthompson** [comment]
@merit-sno this looks awesome, nice work on the quick PR. For testing/linting I've just cloned things locally and done some light poking mostly just for syntax. Do we have more formal test steps to validate the functionality of these changes?

File: setup.py (hunk 5 above)
Comment (merit-schelcj) [reply setup.py]: Since we intend to push this upstream we probably shouldn't touch these for the time being.

File: netsuite/client.py (hunk 1 above)
Comment (merit-sno) [reply netsuite/client.py]: Regarding python 3 compatibility, I have run a version of this code in P3 with no issues.  This structure was provided by a Python 2 developer, which I believe that this code should still support, especially if we are going to create a pull request back to the project owner.

I will examine the error condition report.

File: netsuite/client.py (hunk 2 above)
Comment (merit-sno) [reply netsuite/client.py]: I think this is another issue with P3 / P2 compatibility.  The relevant module for P2 is urlparse, whereas in P3 it is urllib.parse.

File: setup.py (hunk 5 above)
Comment (merit-sno) [reply setup.py]: I'm confused.  I could not find a setup.py kwargs reserved word dictionary for this so I assumed there is not one.  I would appreciate a pointer to such a specification.

Comment (merit-mthompson) [reply setup.py]: https://packaging.python.org/tutorials/packaging-projects/

File: netsuite/client.py (hunk 1 above)
Comment (merit-mthompson) [reply netsuite/client.py]: The original package only lists Python 3.x support. A few places this is demonstrated:

- The little graphic near the header at https://github.com/jmagnusson/netsuite
- Tests which only test for 3.x at https://github.com/jmagnusson/netsuite/blob/master/.travis.yml
- The package information with only lists 3.x at https://github.com/jmagnusson/netsuite/blob/master/setup.py#L36

_(We also have a PSWD suggestion that all new code be Python 3: https://confluence.merit.edu/display/SD/Python)_

File: netsuite/client.py (hunk 2 above)
Comment (merit-mthompson) [reply netsuite/client.py]: The original author was only coding and testing for Python 3. See https://github.com/merit-network/netsuite/pull/2#discussion_r388426592

File: netsuite/client.py [hunk 6]
```diff
@@ -83,8 +85,54 @@ def wrapper(self, *args, **kw):
...
+        """
+        Assign the dynamic host domain component to a class variable
+        """
+        wsdl_url = kwargs['wsdl_url']
+        kwargs.pop('wsdl_url',None)
+
+        parsed_wsdl_url = urlparse(wsdl_url)
+        self._wsdl_url = f'{parsed_wsdl_url.scheme}://{parsed_wsdl_url.netloc}/'
+
+        super().__init__(**kwargs)
```
Comment (merit-mthompson) [inline netsuite/client.py]: Could this maybe turn in to:

```python

    def __init__(self, wsdl_url, *args, **kwargs):
        """
        Assign the dynamic host domain component to a class variable
        """
        parsed_wsdl_url = urlparse(wsdl_url)
        self._wsdl_url = f'{parsed_wsdl_url.scheme}://{parsed_wsdl_url.netloc}/'

        super().__init__(*args, **kwargs)
```

It makes the init() requirement of wsdl_url more explicit and saves you the trouble of having to manipulate args/kwargs for the super(). (Generally, kwargs are optional arguments while positional args are required. It's also not very common in Python to pop kwargs like that.)

Then when you instantiate the class you remove the key and pass the URL as a positional argument.

```python
    return NetSuiteTransport(
        self._generate_wsdl_url(),
        session=self.session,
        cache=self.cache,
    )
```

**Review by merit-mthompson — CHANGES_REQUESTED** [review]
Would like to see the __init__ follow Python-isms a bit more closely. Otherwise, looks great. I'm not trying to be a weenie about how we do things, just trying to follow best practices.

File: netsuite/client.py (hunk 6 above)
Comment (merit-mthompson) [reply netsuite/client.py]: And for a deeper-dive into args/kwargs: https://www.digitalocean.com/community/tutorials/how-to-use-args-and-kwargs-in-python-3

Comment (merit-sno) [reply netsuite/client.py]: It seems like bad form to mix arguments by position, generic list, and key/value all in the same call.

**Review by merit-mthompson — APPROVED** [review]

**Review by merit-schelcj — CHANGES_REQUESTED** [review]
Looks good other than changing the author details. Leave the original authors info intact. If these changes aren't accepted upstream then we update this info.

**Review by merit-schelcj — APPROVED** [review]

**Pull request merged**